src/utils.ts

```diff
@@ -495,3 +499,76 @@ export const defaultConvertTaskResultForConsoleTable: ConsoleTableConverter = (
   }
   /* eslint-enable perfectionist/sort-objects */
 }
+
+interface WithConcurrencyOptions<R> {
+  fn: () => Promise<R>
+  iterations: number
+  limit: number
+  now?: () => number
+  signal?: AbortSignal
+  time?: number
+}
+
+/**
+ * Creates a concurrency limiter that can execute functions with a maximum concurrency limit.
+ * @param options - The resource containing the function to execute and other options
+ * @returns A promise that resolves to an array of results.
+ */
+export const withConcurrency = async <R>(options: WithConcurrencyOptions<R>): Promise<R[]> => {
+  const { fn, iterations, limit, now = performanceNow, signal, time = 0 } = options
+
+  const maxWorkers = iterations === 0 ? limit : Math.max(0, Math.min(limit, iterations))
+
+  const errors: Error[] = []
+  const results: R[] = []
+
+  let finished = false
+  let nextIndex = 0
+  const startTime = Number.isFinite(time) && time !== 0 ? now() : 0
+
+  const doNext = (): boolean => {
+    if (finished) return false
+    if (iterations !== 0 && nextIndex >= iterations) return false
+    nextIndex++
+    return true
+  }
+
+  const pushResult = (r: R) => { if (!finished) results.push(r) }
+  const pushError = (e: unknown) => { errors.push(toError(e)) }
+
+  const onAbort = () => { finished = true }
+
+  if (signal) {
+    if (signal.aborted) return []
+    signal.addEventListener('abort', onAbort)
+  }
+
+  const worker = async () => {
+    while (!finished) {
```

Comment:
the caller is blocked until there's a room to proceed the next. I think it's fine for tinybench usage and avoid the usage of intermediate mechanism to store internally the work left.